File: beam/lib/bowser_brain/chrome.ex
defmodule BowserBrain.Chrome do
  @moduledoc """
  Browser-chrome surface points (Mod API v1). Buttons appear in every
  window's toolbar; clicks come back as events:

      Chrome.add_button("reader", "Reader", symbol: "book")
      # ... then in a mod:
      def handle_event(%{"event" => "chrome_click", "id" => "reader"}, state), do: ...

  Omnibar input starting with `:` arrives as
  `%{"event" => "omnibar_command", "text" => "..."}` — mods own that namespace.
  """

  alias BowserBrain.Bridge

  @doc """
  Add/update a native bar, reserving webpage space on edge: :top/:bottom/:left/:right.
  size: 16..200 points (height for horizontal bars, width for vertical bars).
  view is a View DSL tree; style accepts background/foreground/border #rrggbb.
  Events arrive as surface events with surface: "toolbar:<id>" and webview.
  Call in init_mod and mod_reloaded. Bars die with the calling mod, replay on
  reconnect, and appear in new browser windows. Same ids shadow older owners.
  """
  def put_toolbar(id, view, opts \\ []), do: BowserBrain.Toolbars.put(to_string(id), view, opts)
  def remove_toolbar(id), do: BowserBrain.Toolbars.remove(to_string(id))
  def toolbars, do: BowserBrain.Toolbars.list()

  @doc """
  Apply a native browser skin owned by the calling mod process. Accepts a map
  with hex #rrggbb colors: background, foreground, button_background,
  button_foreground, accent, border; button_style: "flat" or "beveled";
  show_navigation: boolean; title_size: 9..16; corner_radius: 0..12.
  Full-window inner outline: window_border: #rrggbb, window_border_width:
  0..12 points, window_border_style: "flat" or "beveled". Does not alter
  macOS window shape or shadow.
  Atom or string keys are accepted. Omitted properties use native defaults.
  Call in init_mod and on mod_reloaded; reconnects replay automatically.
  Disabling/deleting the mod restores the previous theme, or native defaults.
  Returns :ok or {:error, :invalid_theme}; invalid themes leave state intact.
  """
  def set_theme(theme), do: BowserBrain.ShellTheme.set(theme)

  @doc "Remove only the calling process's theme, restoring the previous skin."
  def reset_theme, do: BowserBrain.ShellTheme.reset()

  @doc "Current effective native shell theme (string-keyed map)."
  def theme, do: BowserBrain.ShellTheme.current()

  @doc "Add (or replace) a toolbar button. `symbol:` is an SF Symbol name."
  def add_button(id, title, opts \\ []) do
    Bridge.cast_msg(%{
      op: "chrome",
      chrome: "add_button",
      id: id,
      title: title,
      symbol: Keyword.get(opts, :symbol)
    })
  end

  @doc "Remove a toolbar button by id."
  def remove_button(id) do
    Bridge.cast_msg(%{op: "chrome", chrome: "remove_button", id: id})
  end

  @doc """
  Add (or replace) an item in the native View menu. `key:` is an optional
  ⌘-key equivalent (single character). Clicks arrive as `chrome_click`
  events with this id — same contract as buttons. Shell state: re-assert
  on "hello".
  """
  def add_menu_item(id, title, opts \\ []) do
    Bridge.cast_msg(%{
      op: "chrome",
      chrome: "add_menu_item",
      id: id,
      title: title,
      key: Keyword.get(opts, :key),
      # nil = plain action item; true/false = checkable toggle.
      checked: Keyword.get(opts, :checked)
    })
  end

  @doc "Remove a View-menu item by id."
  def remove_menu_item(id) do
    Bridge.cast_msg(%{op: "chrome", chrome: "remove_menu_item", id: id})
  end

  @doc """
  Register an omnibar command for visual recognition: while the user types
  `:name …`, the omnibar shows `hint`. Re-register on "hello" (shell state
  dies with the engine).
  """
  def register_command(name, hint) do
    Bridge.cast_msg(%{op: "chrome", chrome: "register_command", name: name, hint: hint})
  end

  @doc """
  Open a tab. `activate: true` switches to it; by default the webview is
  created detached — it exists, loads, and shows up in tab UI, but what the
  user is looking at doesn't move. Switch later with `Surface.activate_tab/1`.
  """
  def open_tab(url \\ nil, opts \\ []) do
    msg = %{
      op: "chrome",
      chrome: "open_tab",
      url: url,
      activate: Keyword.get(opts, :activate, false)
    }

    # profile: the tab goes to (or creates) a window of that profile; tabs
    # never cross profiles.
    msg = if p = Keyword.get(opts, :profile), do: Map.put(msg, :profile, p), else: msg
    Bridge.cast_msg(msg)
  end

  @doc "A new browser window bound to a profile (see BowserBrain.Profiles)."
  def open_window(profile_id) do
    Bridge.cast_msg(%{op: "chrome", chrome: "open_window", profile: profile_id})
  end

  @doc """
  Deprecated no-ops (bowser-browser-cdd): there is no native tab bar any
  more. One window holds N in-memory webviews and tab UI is entirely a mod's
  job. Kept so mods written against the old API keep running.
  """
  def hide_tab_bar, do: Bridge.cast_msg(%{op: "chrome", chrome: "hide_tab_bar"})

  @doc false
  def show_tab_bar, do: Bridge.cast_msg(%{op: "chrome", chrome: "show_tab_bar"})
end

File: shell/Sources/Bowser/Chrome.swift
import AppKit
import Foundation

/// Mod-driven chrome surface points (Mod API v1): the brain sends "chrome"
/// ops, the shell applies them to every browser window. State lives here so
/// windows created later inherit the current surface.
@MainActor
enum ChromeSurface {
    struct ModButton: Equatable {
        let id: String
        let title: String
        let symbol: String?
    }

    struct ModMenuItem: Equatable {
        let id: String
        let title: String
        let key: String?
        /// nil = plain action item; true/false = stateful toggle with checkmark.
        let checked: Bool?
    }

    private static var profileButtons: [String: [ModButton]] = [:]
    private static var profileMenus: [String: [ModMenuItem]] = [:]
    private static var profileThemes: [String: ShellTheme] = [:]
    private static var profileToolbars: [String: [ModToolbar]] = [:]
    private static var profileCommands: [String: [String: String]] = [:]
    static func buttons(for profile: String) -> [ModButton] { buttons + (profileButtons[profile] ?? []) }
    static func menus(for profile: String) -> [ModMenuItem] { menuItems + (profileMenus[profile] ?? []) }
    static func theme(for profile: String) -> ShellTheme { profileThemes[profile] ?? theme }
    static func toolbars(for profile: String) -> [ModToolbar] { toolbars + (profileToolbars[profile] ?? []) }
    static func commands(for profile: String) -> [String: String] { commands.merging(profileCommands[profile] ?? [:]) { _, new in new } }

    private(set) static var buttons: [ModButton] = []
    /// Mod-owned entries appended to the native View menu; clicks come back
    /// as chrome_click, exactly like band buttons.
    private(set) static var menuItems: [ModMenuItem] = []
    /// Registered omnibar commands: name → hint (shown while typing).
    private(set) static var commands: [String: String] = [:]
    private(set) static var theme: ShellTheme = .native
    private(set) static var toolbars: [ModToolbar] = []
    private static var controllers: [ObjectIdentifier: BrowserWindowController] = [:]

    static func register(_ controller: BrowserWindowController) {
        controllers[ObjectIdentifier(controller)] = controller
        controller.syncModButtons()
        controller.syncShellTheme()
        controller.syncToolbars()
    }

    static func unregister(_ controller: BrowserWindowController) {
        controllers.removeValue(forKey: ObjectIdentifier(controller))
    }

    static func handle(_ object: [String: Any]) {
        guard let action = object["chrome"] as? String else {
            NSLog("Bowser: chrome op without action")
            return
        }

        let profile = object["profile"] as? String
        switch action {
        case "set_toolbars":
            guard let json = object["toolbars"] as? [[String: Any]], json.count <= 16 else { return }
            let parsed = json.compactMap(ModToolbar.init(json:))
            guard parsed.count == json.count, Set(parsed.map(\.id)).count == parsed.count else { return }
            if let profile { profileToolbars[profile] = parsed } else { toolbars = parsed }
            for controller in controllers.values { controller.syncToolbars() }
            return
        case "set_theme":
            guard let json = object["theme"] as? [String: Any],
                  let next = ShellTheme(json: json) else { return }
            if let profile { profileThemes[profile] = next } else { theme = next }
            for controller in controllers.values { controller.syncShellTheme() }
            return
        case "add_button":
            guard let id = object["id"] as? String else { return }
            var entries = profile.map { profileButtons[$0] ?? [] } ?? buttons
            entries.removeAll { $0.id == id }
            entries.append(ModButton(
                id: id,
                title: object["title"] as? String ?? id,
                symbol: object["symbol"] as? String
            ))
        if let profile { profileButtons[profile] = entries } else { buttons = entries }
        case "remove_button":
            guard let id = object["id"] as? String else { return }
            if let profile { profileButtons[profile]?.removeAll { $0.id == id } } else { buttons.removeAll { $0.id == id } }
        case "add_menu_item":
            guard let id = object["id"] as? String else { return }
            var entries = profile.map { profileMenus[$0] ?? [] } ?? menuItems
            entries.removeAll { $0.id == id }
            entries.append(ModMenuItem(
                id: id,
                title: object["title"] as? String ?? id,
                key: object["key"] as? String,
                checked: object["checked"] as? Bool
            ))
            if let profile { profileMenus[profile] = entries } else { menuItems = entries }
            // NSApplication.shared, not NSApp: NSApp is nil in headless tests.
            (NSApplication.shared.delegate as? AppDelegate)?.rebuildModMenuItems()
            return
        case "remove_menu_item":
            guard let id = object["id"] as? String else { return }
            if let profile { profileMenus[profile]?.removeAll { $0.id == id } } else { menuItems.removeAll { $0.id == id } }
            (NSApplication.shared.delegate as? AppDelegate)?.rebuildModMenuItems()
            return
        case "register_command":
            guard let name = object["name"] as? String else { return }
            if let profile { profileCommands[profile, default: [:]][name] = object["hint"] as? String ?? name }
            else { commands[name] = object["hint"] as? String ?? name }
            return
        case "hide_tab_bar", "show_tab_bar":
            // Accepted and ignored: there IS no native tab bar any more
            // (bowser-browser-cdd). Mods written against the old API — the
            // dock calls hide_tab_bar on every hello — must not break.
            return
        case "open_tab":
            guard let delegate = NSApp.delegate as? AppDelegate else { return }
            // Background by default: open_tab creates the webview, the dock
            // (or an explicit activate_tab) decides what the user sees.
            delegate.openTab(
                url: object["url"] as? String,
                activate: object["activate"] as? Bool ?? false,
                profile: object["profile"] as? String
            )
            return
        case "open_window":
            guard let delegate = NSApp.delegate as? AppDelegate else { return }
            let controller = delegate.openWindow(profile: Profile.find(object["profile"] as? String))
            controller.window?.makeKeyAndOrderFront(nil)
            NSApp.activate()
            return
        default:
            NSLog("Bowser: unknown chrome op \(action)")
            return
        }
        for controller in controllers.values {
            controller.syncModButtons()
        }
    }

    static func emit(_ message: [String: Any]) {
        BrainBridge.shared.send(message)
    }
}
